File: backend/src/graph/nodes.ts
import { sql_maker } from "../tools/sql_maker";
import { validator_tool } from "../tools/validator.tool";
import { sql_executor } from "../tools/sql_executor";
import { GraphState } from "./state";
import { check_relevance_tool } from "../tools/relevance_checker";

export const generate_sql_node = async (state: typeof GraphState.State) => {
  const { question, feedback, schema, restrictedColumns } = state;
  // Determine if the question is answerable by the schema
  const isRelevant = await check_relevance_tool(question, schema);
  if (!isRelevant) {
    console.warn(
      "Relevance Check Failed: Question deemed unrelated to schema. Proceeding anyway..."
    );
    // throw new Error("Invalid Request: The question is unrelated to the provided database schema.");
  }

  let debugInfo = "";
  // --- Relaxed Column Selection ---
  // We simply pass the restrictedColumns (now acting as "Selected Columns")
  // to the sql_maker to guide the SELECT clause.

  let sqlOrError = await sql_maker(
    question,
    schema,
    feedback || "",
    restrictedColumns || []
  );

  // Check if result is the JSON error (legacy check, just in case)
  try {
    const parsed = JSON.parse(sqlOrError);
    if (parsed.error && parsed.missingColumns) {
      // Should not happen now, but keeping for safety
      return {
        sql: sqlOrError,
        iterations: (state.iterations || 0) + 1,
      };
    }
  } catch (e) {
    // Not JSON, normal SQL
  }

  // Check if result is the JSON error
  try {
    const parsed = JSON.parse(sqlOrError);
    if (parsed.error && parsed.missingColumns) {
      // It's a missing column error.
      // We can throw or separate it. Ideally, we return it as "result" or a special state to stop execution.
      // But strict graph expects "sql".
      // Let's wrap it in a special "ERROR:" string if simple, or better:
      // The graph expects a string for 'sql'. If we return JSON, the validator might fail.
      // Let's assume valid=false and pass this as feedback?
      // Or better, we need the frontend to receive this.

      // Hack: If it's strict error, return it as the "sql" value but marked.
      // Actually, we can return it as is. The validator will fail syntax check (it's JSON not SQL).
      // But we want to fail gracefully.
      return {
        sql: sqlOrError, // Pass JSON string as SQL
        iterations: (state.iterations || 0) + 1,
      };
    }
  } catch (e) {
    // Not JSON, so it's likely SQL. Continue.
  }

  return {
    sql: sqlOrError,
    iterations: (state.iterations || 0) + 1,
  };
};

export const validate_sql_node = async (state: typeof GraphState.State) => {
  const { sql, schema } = state;

  // --- SECURITY CHECK (Proposed Restriction) ---

  // const forbiddenKeywords = ["DELETE", "UPDATE", "ALTER", "DROP", "TRUNCATE", "INSERT"];
  // const upperSQL = sql.toUpperCase();
  // if (forbiddenKeywords.some(keyword => upperSQL.includes(keyword))) {
  //     return {
  //         valid: false,
  //         feedback: "Invalid Query: Modifying the database (DELETE, UPDATE, ALTER, etc.) is not allowed. Read-only queries only."
  //     };
  // }

  // ---------------------------------------------
  const validation = await validator_tool(sql, schema);
  return {
    valid: validation.valid,
    feedback: validation.reasoning,
  };
};

export const execute_sql_node = async (state: typeof GraphState.State) => {
  const { sql, db } = state;
  try {
    const result = await sql_executor(sql, db);
    return { result };
  } catch (error: any) {
    return {
      valid: false,
      feedback: `Runtime Error: ${error.message}. Please fix the SQL.`,
      iterations: (state.iterations || 0) + 1,
    };
  }
};

File: backend/app/graph/nodes.py
from app.graph.state import GraphState
from app.services.llm_service import get_llm
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
import sqlite3
import pandas as pd
import json

llm = get_llm()

# --- Prompts ---
ROUTER_PROMPT = """You are analyzing an intent for a database querying system.
Given the user's question, decide if they want:
1. "data" - A tabular response containing actual raw data pulled via an SQL query (e.g., "show me the top 10 rows", "give me first 1000", "filter by x", "get all the").
2. "chat" - A text-based chatbot response, such as a general question, a request for a "summary of the table" or insights, an explanation, or simple conversational chat.

Question: {question}

Return ONLY a valid JSON object matching this schema:
{{"intent": "data" | "chat"}}
"""

GENERATE_SQL_PROMPT = """
You are a SQL Expert. Given an input question and a database schema, generate a valid SQLite query.
Schema:
{schema}

Question: {question}

Restricted Columns (Select only these if not empty): {restricted_columns}
Previous Feedback (if any): {feedback}

Return ONLY the raw SQL query. Do not wrap in markdown or code blocks. Unless the user specifies a limit, use a default LIMIT 1000 to ensure enough data for visualization.
"""

VALIDATE_SQL_PROMPT = """
Check if the following SQL query is valid for SQLite and safe (Read-Only).
Schema:
{schema}

Query: {sql}

Return a JSON object: {{"valid": boolean, "reasoning": "string"}}
"""

CHAT_PROMPT = """You are a helpful Data Analysis Chatbot.
Based on the following database schema and a preview of the data:
Schema:
{schema}

Preview of first 5 rows:
{preview}

Answer the user's question or summarize the table as requested. Provide a concise, clear text response. Do NOT provide tabular data if they want a summary, just describe it or answer their text question.
Question: {question}
"""

# --- Nodes ---

async def route_intent_node(state: GraphState):
    prompt = ChatPromptTemplate.from_template(ROUTER_PROMPT)
    chain = prompt | llm | StrOutputParser()
    response = await chain.ainvoke({"question": state["question"]})
    try:
        cleaned = response.replace("```json", "").replace("```", "").strip()
        data = json.loads(cleaned)
        intent = data.get("intent", "data")
    except:
        intent = "data"
        
    return {"intent": intent}
    
async def chat_node(state: GraphState):
    db_path = state["db_path"]
    
    preview_data = {}
    if db_path:
        try:
            conn = sqlite3.connect(db_path)
            cursor = conn.cursor()
            cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
            tables = [r[0] for r in cursor.fetchall()]
            for t in tables:
                df = pd.read_sql_query(f'SELECT * FROM "{t}" LIMIT 5', conn)
                df = df.replace([float('inf'), float('-inf'), float('nan')], None)
                preview_data[t] = df.to_dict(orient="records")
            conn.close()
        except:
            pass

    prompt = ChatPromptTemplate.from_template(CHAT_PROMPT)
    chain = prompt | llm | StrOutputParser()
    response = await chain.ainvoke({
        "schema": state["schema"],
        "preview": json.dumps(preview_data, default=str),
        "question": state["question"]
    })
    
    return {"result": response, "valid": True, "intent": "chat"}

async def generate_sql_node(state: GraphState):
    prompt = ChatPromptTemplate.from_template(GENERATE_SQL_PROMPT)
    chain = prompt | llm | StrOutputParser()
    
    response = await chain.ainvoke({
        "schema": state["schema"],
        "question": state["question"],
        "restricted_columns": state.get("restricted_columns", []),
        "feedback": state.get("feedback", "")
    })
    
    # Clean response
    sql = response.replace("```sql", "").replace("```", "").strip()
    
    return {"sql": sql, "iterations": state["iterations"] + 1}

async def validate_sql_node(state: GraphState):
    prompt = ChatPromptTemplate.from_template(VALIDATE_SQL_PROMPT)
    chain = prompt | llm | StrOutputParser()
    
    response = await chain.ainvoke({
        "schema": state["schema"],
        "sql": state["sql"]
    })
    
    try:
        # Clean response
        cleaned = response.replace("```json", "").replace("```", "").strip()
        data = json.loads(cleaned)
        return {"valid": data["valid"], "feedback": data["reasoning"]}
    except:
        # If parsing fails, assume valid but warn? Or strictly fail.
        # Fallback for robustness
        return {"valid": True, "feedback": "Validation parsing failed, proceeding cautiously."}

async def execute_sql_node(state: GraphState):
    db_path = state["db_path"]
    sql = state["sql"]
    
    if not db_path:
         return {"valid": False, "feedback": "Database path missing."}

    try:
        conn = sqlite3.connect(db_path)
        # Security: Allow read-only? 
        if "DROP" in sql.upper() or "DELETE" in sql.upper() or "UPDATE" in sql.upper():
             return {"valid": False, "feedback": "Write operations are not allowed."}

        df = pd.read_sql_query(sql, conn)
        df = df.replace([float('inf'), float('-inf'), float('nan')], None)
        result = df.to_dict(orient="records")
        conn.close()
        
        return {"result": result, "valid": True}
    except Exception as e:
        return {"valid": False, "feedback": f"Runtime Error: {str(e)}"}
